File: 2026-03/2026-03-13/parking_fee_calculator.py
from datetime import time
import math

def calculate_parking_fee(park_time, pickup_time):
    park_time_object = time.fromisoformat(park_time)
    pickup_time_object = time.fromisoformat(pickup_time)
    hourly_fee = 3
    overnight_fee = 10
    minimum_fee = 5

    park_time_minutes_since_midnight = park_time_object.hour * 60 + park_time_object.minute
    pickup_time_minutes_since_midnight = pickup_time_object.hour * 60 + pickup_time_object.minute
    parking_duration_minutes = pickup_time_minutes_since_midnight - park_time_minutes_since_midnight
    overnight_parking = True if parking_duration_minutes < 0 else False

    parking_duration_minutes = parking_duration_minutes % 1440
    parking_hours = math.ceil(parking_duration_minutes / 60)

    parking_fee = parking_hours * hourly_fee
    parking_fee = parking_fee + overnight_fee if overnight_parking else parking_fee
    parking_fee = minimum_fee if parking_fee < minimum_fee else parking_fee

    return f"${parking_fee}"

# --- TEST SUITE ---

tests_text = r'''
1. calculate_parking_fee("09:00", "11:00") should return "$6".
2. calculate_parking_fee("10:00", "10:30") should return "$5".
3. calculate_parking_fee("08:10", "10:45") should return "$9".
4. calculate_parking_fee("14:40", "23:10") should return "$27".
5. calculate_parking_fee("18:15", "01:30") should return "$34".
6. calculate_parking_fee("11:11", "11:10") should return "$82".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 03 March/03-13/parkingFeeCalculator.js
function calculateParkingFee(parkTime, pickupTime) {
  const parkTimeTemporal = Temporal.PlainTime.from(parkTime);
  const pickupTimeTemporal = Temporal.PlainTime.from(pickupTime);
  const hourlyFee = 3;
  const overnightFee = 10;
  const minimumFee = 5;
  let parkingHours;

  const overnightParking = Temporal.PlainTime.compare(parkTimeTemporal, pickupTimeTemporal) === 1 ? true : false;
  const parkingDuration = pickupTimeTemporal.since(parkTimeTemporal);
  
  if (overnightParking) {
    parkingHours = 24 + parkingDuration.hours;
  }
  else {
    parkingHours = parkingDuration.hours;
    
    if (parkingDuration.minutes !== 0)
      ++parkingHours;
  }

  let parkingFee = parkingHours * hourlyFee;
  parkingFee = overnightParking ? parkingFee + overnightFee : parkingFee;
  parkingFee = parkingFee < minimumFee ? minimumFee : parkingFee;

  return `$${parkingFee}`;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. calculateParkingFee("09:00", "11:00") should return "$6".
2. calculateParkingFee("10:00", "10:30") should return "$5".
3. calculateParkingFee("08:10", "10:45") should return "$9".
4. calculateParkingFee("14:40", "23:10") should return "$27".
5. calculateParkingFee("18:15", "01:30") should return "$34".
6. calculateParkingFee("11:11", "11:10") should return "$82".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);